22.0.0.10 post (#2575)
OpenLiberty/blogs: posts/2022-09-27-22.0.0.10.adoc

View the list of fixed bugs in link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A220010+label%3A%22release+bug%22[22.0.0.10].

[#run]
== Run your apps using 22.0.0.10

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
    <groupId>io.openliberty</groupId>
    <artifactId>openliberty-runtime</artifactId>
    <version>22.0.0.10</version>
    <type>zip</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty', name: 'openliberty-runtime', version: '[22.0.0.10,)'
}
----

Or if you're using Docker:

[source]
----
FROM open-liberty
----

Or take a look at our link:{url-prefix}/downloads/[Downloads page].

[link=https://stackoverflow.com/tags/open-liberty]
image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]

// https://github.com/OpenLiberty/open-liberty/issues/21783
[#oidc]
== Use issuer claim for selecting which configuration to use for a JWT inbound propagation request (OpenID Connect Client 1.0)

> Starting in 22.0.0.10, the OpenID Connect Client 1.0 feature supports using the issuer claim from the JWT or JWS access token for selecting which `openidConnectClient` configuration to use for a JWT inbound propagation request.  Prior to this functionality, complicated authentication filters were required if more than one issuer was used for the same resource. Instead, the OpenID Connect Client 1.0 feature can now be configured with simpler `openidConnectClient` elements for each required issuer.

Reviewer: ```suggestion
Starting in 22.0.0.10, the link:{url-prefix}/docs/latest/reference/feature/openidConnectClient-1.0.html[OpenID Connect Client 1.0] feature supports using the issuer claim from the JWT or JWS access token for selecting which `openidConnectClient` configuration to use for a JWT inbound propagation request.  Prior to this release, complicated authentication filters were required if more than one issuer was used for the same resource. Instead, the OpenID Connect Client 1.0 feature can now be configured with simpler `openidConnectClient` elements for each required issuer.
```

A Liberty OpenID Connect Relying Party will automatically use the issuer claim from a JWT or JWS access token to select the `openidConnectClient` configuration with a matching `issuer` attribute. 

> For example, for an issuer claim value of `"https://hostname/op1"` and a `server.xml` file configured as follows:

Reviewer: ```suggestion
The following `server.xml` file example shows two Open ID connect client configurations, with an link:{url-prefix}/docs/latest/authentication-filters.html[authentication filter] configured to route requests to the `RP2` configuration. If a request with an issuer claim value of `https://hostname/op1` is presented, the `RP1` configuration is selected unless the `rp2filter` applies to the request.
```

[source, xml]
----
<server>
    <featureManager>
        <feature>openidConnectClient-1.0</feature>
    </featureManager>
    ...
    <openidConnectClient id="RP1" issuer="https://hostname/op1"  ... />
    <openidConnectClient id="RP2" issuer="https://hostname/op2" authFilterRef="rp2filter" .../>
    <authFilter id="rp2filter">
    ...
    </authFilter>
    ...
</server>
----

> The `"RP1"` configuration is selected as long as the `"rp2filter"` does not apply to the request. Therefore, an attempt to find which `openidConnectClient` configuration applies to the request will be done using the following algorithm:

Reviewer: ```suggestion
Open Liberty selects the `openidConnectClient` configuration to use for a JWT request according to the following algorithm:
```

> 1. If the auth filter for a `openidConnectClient` matches the request, choose that configuration.

Reviewer: ```suggestion
1. If the authentication filter for a `openidConnectClient` configuration matches the request, choose that configuration.
```

> 2. Check the issuer from the JWT against the issuers of all config elements. If the issuer from the JWT matches a single `openidConnectClient` configuration element, choose that configuration.

Reviewer: ```suggestion
2. Check the issuer claim from the JWT against the `issuer` attributes from of all `openidConnectClient` elements. If the issuer claim from the JWT matches the `issuer` attribute in only one `openidConnectClient` configuration element, choose that configuration.
```

> 3. If the issuer from the JWT matches the issuer of multiple `openidConnectClient` configuration elements, choose the first configuration matched.

Reviewer: ```suggestion
3. If the issuer claim from the JWT matches the `issuer` attribute in multiple `openidConnectClient` configuration elements, choose the first configuration that matches.
```

> 4. If the issuer does not match the issuer of any configuration elements, choose between all of the config elements in a non-deterministic manner.

Reviewer: ```suggestion
4. If the issuer claim from the JWT does not match the `issuer` attribute of any `openidConnectClient` configuration elements, choose one of the configuration elements in a nondeterministic manner.
```

Reviewer: I feel like it's unclear what happens in the case of "4"- does it validate the request based on other criteria than the issuer? Does it continue checking configs until it finds a valid option, or exhausts the choices?

Maybe more detail than is needed here but something I want to sort out for the docs

Author: The way I read this is that if 1-3 do not apply, then 'randomly' pick any of the openidConnectClient config elements.

@teddyjtorres can you expand on 4.  Is this nondeterministic per request, or is it nondeterministic for the first request, and then that result is applied to all following requests until server restarts (or I suppose until the feature is unloaded and reloaded)?

Reviewer 2: Hi. Yes, it is non-deterministic (as it's always been prior to this change) since we cannot guarantee the order in which config admin will order these configs. It might not be as it appears in the server.xml. We are just stating what it always has been.

Reviewer 2: Therefore, to avoid confusion, the following is proposed, since this is what it always has done,

If the issuer claim from the JWT does not match the `issuer` attribute of any `openidConnectClient` configuration elements, choose the first configuration that matched.

Reviewer: @teddyjtorres - to clarify- what does matched mean in the case of 4?

> If the issuer claim from the JWT _does not match_ the issuer attribute of any openidConnectClient configuration elements, choose the first configuration that _matched_.

For the steps 1-3, "match" seems to mean that the issuer claim from the JWT matches the `issuer` attribute of one or more `openidConnectClient` configuration elements. But for step 4 we are saying `...the issuer claim from the JWT does not match the issuer attribute of any openidConnectClient configuration elements`- so what is it that "matches" in `the first configuration that matched.` ?

> For more information about OpenID Connect Client refer to the link:https://openid.net/specs/openid-connect-core-1_0.html[OpenID Connect Client specification].

Reviewer: ```suggestion
For more information about OpenID Connect Client, refer to the link:https://openid.net/specs/openid-connect-core-1_0.html[OpenID Connect Client specification].
```

[#CVEs]
== Security vulnerability (CVE) fixes in this release
[cols="5*"]
|===
|CVE |CVSS Score |Vulnerability Assessment |Versions Affected |Version Fixed |Notes

|http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2022-34165[CVE-2022-34165]
|5.4
|HTTP header injection
|17.0.0.3 - 22.0.0.9
|22.0.0.10
|link:https://github.com/OpenLiberty/open-liberty/issues/22425[See issue #2245]
|===

For a list of past security vulnerability fixes, reference the link:{url-prefix}/docs/latest/security-vulnerabilities.html[Security vulnerability (CVE) list].

[#bugs]
== Notable bugs fixed in this release

We’ve spent some time fixing bugs. The following sections describe just some of the issues resolved in this release. If you’re interested, here’s the  link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A220010+label%3A%22release+bug%22[full list of bugs fixed in 22.0.0.10].

* link:https://github.com/OpenLiberty/open-liberty/issues/20599[JDBC connection not validated when numConnectionsPerThreadLocal is used]
+

> When a JDBC datasource is configured with `validationTimeout` the connection pool should validate a connection before it is passed back to the application. If the connection is invalid (e.g. after a database failover event), other connections in the pool should be purged according to the `purgePolicy` configuration. The application should therefore rarely receive an invalid database connection.

Reviewer: ```suggestion
When a JDBC datasource is configured with `validationTimeout`, the connection pool should validate a connection before it is passed back to the application. If the connection is invalid (e.g. after a database failover event), other connections in the pool should be purged according to the `purgePolicy` configuration. The application should therefore rarely receive an invalid database connection.
```

+

> However, if the connection manager is configured with `numConnectionsPerThreadLocal=1`, once a connection is stashed as a `ThreadLocal`, it isn't validated before being used by the application resulting in a `SQLException`. Also, if the application calls `isValid` on the connection and the connection is determined to be invalid, the connection manager does not apply the `purgePolicy` logic to other connections in the pool, so clients may see multiple errors until each of the worker threads is used. 

Reviewer: ```suggestion
However, if the connection manager is configured with `numConnectionsPerThreadLocal=1`, once a connection is stashed as a `ThreadLocal`, it isn't validated before being used by the application, which results in a `SQLException`. Also, if the application calls `isValid` on the connection and the connection is determined to be invalid, the connection manager does not apply the `purgePolicy` logic to other connections in the pool, so clients may see multiple errors until each of the worker threads is used. 
```
